doc/vim-go.txt

```diff
@@ -1815,6 +1829,200 @@ The `gohtmltmpl` filetype is automatically set for `*.tmpl` files; the
 `gotexttmpl` is never automatically set and needs to be set manually.
 
 
+==============================================================================
+DEBUGGER                                                            *go-debug*
+
+Vim-go comes with a special "debugger mode". This starts a `dlv` process in
+the background and provides various commands to communicate with it.
+
+This debugger is similar to Visual Studio or Eclipse and has the following
+features:
+
+  * Show stack trace and jumps.
+  * List local variables.
+  * List function arguments.
+  * Expand values of struct or array/slice.
+  * Show balloon on the symbol.
+  * Show output of stdout/stderr.
+  * Toggle breakpoint.
+  * Stack operation continue/next/step out.
+
+This feature requires Vim 8.0.0087 or newer with the |+job| feature. Neovim
+does _not_ work (yet).
+This requires a fairly recent version of the Delve debugger (Nov 28 or newer).
+Use `|:GoUpdateBinaries| to install or update it.
+
+                                                              *go-debug-intro*
+GETTING STARTED WITH THE DEBUGGER~
+
+Use |:GoDebugStart| to start the debugger. The first argument is the package
+name, and any arguments after that will be passed on to the program; for
+example:
+>
+    :GoDebugStart . -someflag value
+<
+This may take few seconds. After the code is compiled you'll see three new
+windows: the stack trace on left side, the variable list on the bottom-left,
+and program output at the bottom.
+
+You can add breakpoints with |:GoDebugBreakpoint| (<F9>) and run your program
+with |:GoDebugContinue| (<F5>).
+
+The program will halt on the breakpoint, at which point you can inspect the
+program state. You can go to the next line with |:GoDebugNext| (<F10>) or step
+in with |:GoDebugStep| (<F11>).
+
+Struct values are displayed as `{...}`, array/slices as `[4]`. Use <CR> on the
+variable name to expand the values.
+
+When you're done use |:GoDebugStop| to close the debugging windows and halt
+the `dlv` process, or |:GoDebugRestart| to recompile the code.
```

Comment:
How can I clear previously set breakpoints? Even though I've stopped or restarted the server, the breakpoints are still intact. Is this something we can provide to the user? Seems like there is no way to remove all breakpoints right now and you have to manually remove them one by one.

Reply:
Oh it seems like restarting Vim does remove all breakpoints

Reply:
We could add a `:GoDebugReset` command maybe?

Reply:
Yeap definitely, I think this is much needed otherwise one has to close and restart Vim (which is not something we want to do). Also another thing that I've just seen is, when you change or remove the lines, the breakpoint lines will be invalid. Not %100 related to this, but worth adding as well.

Reply:
> when you change or remove the lines, the breakpoint lines will be invalid. Not %100 related to this, but worth adding as well.

How do other debuggers handle this? Move the breakpoint, or just clear it?

Reply:
In VSCode it moves when you add newlines. So it's dynamic to the content of the line. But not sure if it's worth adding that kind of complexity. But one thing I appreciate is the way to disable/enable breakpoints. This allows to temporary run the whole stack without getting blocked. And they also have something to reset all the breakpoints.  So I think we need the following commands:

```
:GoDebugToggleBreakpoint
:GoDebugDisableBreakpoints
:GoDebugEnableBreakpoints
:GoDebugRemoveBreakpoints
```

O think `:GoDebugReset` would be confusing with `:GoDebugRestart`. So don't let us use `:GoDebugReset`, instead let's have these dedicated commands. We can add `:GoDisableBreakpoints` and `:GoEnableBreakpoints` later if it's to much work. Wdyt?

(Here is the Debug menu from VSCode)

<img width="241" alt="screen shot 2018-02-27 at 6 08 20 pm" src="https://user-images.githubusercontent.com/438920/36737195-8e55ed42-1beb-11e8-91f5-df069b4380ee.png">

Reply:
I'm not so sure of adding four new commands, I find it more annoying to use as e.g. `:GoDebug<Tab>` now has many more commands that are pretty much the same.

I think a better way is to use e.g.:

```
:GoDebugBreakpoint 42         " Toggle breakpoint for a line (add/remove)
:GoDebugBreakpoint toggle     " Toggle all
:GoDebugBreakpoint disable    " Disable all
:GoDebugBreakpoint enable     " Enable all 
:GoDebugBreakpoint remove     " Remove all
```

That being said, I think we should make an issue for this after merging this. It definitely should be done one way or the other, but I don't think it's a critical feature to add *right now*.

Same applies to the other issue with the wrong line. It's definitely something that needs to be improved, but not necessarily in this PR. It's already a large PR, and tracking all comments on GitHub is kind of hard.

Reply:
Agree with that. Let's talk about the UI changes later (command and co).